### PR TITLE
Event

### PR DESCRIPTION
**Pull request by florianv**
This is the event implementation I did for the platform a little bit improved.

See README for usage examples.

### CODE REVIEW

**Comment by eddieajau** [comment]
Florian, that's an awesome amount of work. A few questions:
1. Have you considered using a SplPriorityQueue for registering with priority?
2. How does this implementation compare to "best practice" and what other frameworks are doing?
3. Do we need both Event::addArgument and setArgument? Would just setArgument do?
4. Can you explain a bit more about why the delegating dispatcher is used.

Thanks

**Comment by florianv** [comment]
1. Yes I have considered it, but SplPriorityQueue is buggy :
   Elements with the same priority are dequeued, if you add three elements with the same priority, the first one will be sorted first and the others in the reverse order. Also, you cannot remove elements from an SplPriorityQueue, and it's a heap. So the ListenersPriorityQueue provides the expected behaviour for a priority queue :)
2.  I have looked and Zend framework and Symfony also supports priorities + closures as listeners in their implementation. I made minimal interfaces, so you can easily create a very simple event system by implementing EventInterface and DispatcherInterface. Also in term of testing, it's easier to test a listener by mocking an Event (compared to the "old" Joomla event system).
3. addArgument add only if not existing, setArgument will override an existing one, so we might need both.
4. It's to make a Dispatcher Immutable. In the event case it's easy because you can inject all dependencies via constructor so providing only accessors will make it immutable. I didn't find an easy way to inject the listeners in the Dispatcher constructor, it is easier to use the addListener method and add them one by one. So the Delegating dispatcher will proxy a Dispatcher created easily with the add\* method and expose only the triggerEvent method, which will make it immutable.

For the utility, if you make a plugin system based on this package, let's say like the CMS. If your listeners are provided from a database table, you don't want "people" to be able to remove/add listeners in the Dispatcher, so you will construct your dispatcher from the table rows and inject a DelegatingDispatcher in the system.

File: vendor/Joomla/Event/Dispatcher.php [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+	}
+
+	/**
+	 * Count the number of registered event.
+	 *
+	 * @return  integer  The numer of registered events.
+	 *
+	 * @since   1.0
+	 */
+	public function countEvents()
```
Comment (eddieajau) [inline vendor/Joomla/Event/Dispatcher.php:174]: Would it be better for the class to implement `countable`?

Comment (eddieajau) [inline vendor/Joomla/Event/Dispatcher.php:174]: Never mind - there's multiple count methods.

Comment (florianv) [inline vendor/Joomla/Event/Dispatcher.php:174]: yes there is CountListeners too.

File: vendor/Joomla/Event/Priority.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+/**
+ * An enumeration of priorities for event listeners,
+ * that you are encouraged to use when adding them in the Dispatcher.
+ *
+ * @since  1.0
+ */
+final class Priority
+{
+	const MIN = -3;
```
Comment (eddieajau) [inline vendor/Joomla/Event/Priority.php:17]: Maybe:

LOWEST
VERY_LOW
LOW
NORMAL
HIGH
VERY_HIGH
HIGHEST

??

Also, should we space out the values so developers can fit custom priorities between those we define?

Comment (florianv) [inline vendor/Joomla/Event/Priority.php:17]: Yes.

Maybe with MIN and MAX people won't understand that it's possible to have lower/higher priorities.
I just made a few ones around NORMAL which is the default priority in the Dispatcher, but it's possible to specify -1000 or +10000.

**Comment by eddieajau** [comment]
Florian, can you please re-issue this pull request against the "staging" branch? Thanks.

**Comment by dongilbert** [comment]
Also, you'll need to pull from latest staging. We've moved the Joomla files to the `src/` directory, so as is would cause some problems.

**Comment by florianv** [comment]
Done :)

**Pull request closed without merging**